linkerd/system/src/linux.rs

```diff
@@ -39,9 +42,28 @@ pub fn open_fds(pid: pid_t) -> io::Result<u64> {
 }
 
 pub fn max_fds() -> io::Result<Option<u64>> {
-    let limit = pid::limits_self()?.max_open_files;
-    let max_fds = limit.soft.or(limit.hard).map(|max| max as u64);
-    Ok(max_fds)
+    match Process::myself() {
+        Ok(p) => match p.limits() {
+            Ok(limits) => {
+                let limit = limits.max_open_files;
+                match limit.soft_limit {
+                    process::LimitValue::Unlimited => match limit.hard_limit {
+                        process::LimitValue::Unlimited => Ok(None),
+                        process::LimitValue::Value(v) => Ok(Some(v)),
+                    },
+                    process::LimitValue::Value(v) => Ok(Some(v)),
+                }
+            }
```

Comment:
Are we changing the logic here or just the implementation? Also have we done enough research to make sure that this is the most optimised way of doing this as compared to the original 2-liner code?

Reply:
Since we have changed the package to get limits the implementation of it changes. In the previous package the return type of it was in the form of std::Result<T, Err> hence we could easily use '?' operator to handle the error. But with the new package the return type is a custom error and hence we need to handle errors explicitly.